### PR TITLE
Pima-indians lesson corrected

### PR DESCRIPTION
**Pull request by YaroslavMayorov**

### CODE REVIEW

File: Pima Indians Diabetes and Linear Classifier/Conclusion/task.py [hunk 1]
```diff
@@ -2,16 +2,32 @@
 from utils import plot_classification
 
 
+# The function will load data from a text file with a relative path
+# 'fname'
+# and return it as a pair of arrays - one for different features
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Conclusion/task.py]: and returns it as a pair of arrays: features

File: Pima Indians Diabetes and Linear Classifier/Conclusion/task.py [hunk 2]
```diff
@@ -2,16 +2,32 @@
...
+# and the other one for a diabetes presence
 def read_data(fname):
+    # genfromtxt method loads data from a text file, using a delimeter
+    # to distinguish the columns
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Conclusion/task.py]: based on the provided delimiter.

File: Pima Indians Diabetes and Linear Classifier/Conclusion/task.py [hunk 3]
```diff
@@ -2,16 +2,32 @@
...
     data = np.genfromtxt(fname, delimiter=',')
+    # The data is split into the all but the last column for X and
+    # the last one for y respectively
     X, y = data[:, :-1], data[:, -1]
-    X = (X - X.mean(axis=0)) / X.std(axis=0)  # rescale features
+    # The features are rescaled:
+    # X is standardized - all values are centered around mean
+    # with a unit standard deviation. It means if we will calculate mean
+    # and standard deviation of standard scores it will be 0 and 1 respectively.
+    # This procedure is recommended if the data follows normal distribution
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Conclusion/task.py]: This procedure is recommended for data that follows a normal distribution.

File: Pima Indians Diabetes and Linear Classifier/Conclusion/task.py [hunk 4]
```diff
@@ -2,16 +2,32 @@
...
+    X = (X - X.mean(axis=0)) / X.std(axis=0)
+    # A column with '-1' values is added to the left of the X array
+    # It is a pseudo-feature that will allow us to build initial vectors
+    # for all of the cases. It will be essential later
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Conclusion/task.py]: calculations later on.

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/gradient_descent.py [hunk 5]
```diff
@@ -3,30 +3,50 @@
...
 class GradientDescent:
+    # The weighs will be randomly initiated as the first step of the fitting process.
+    # The default loss function is sigmoid, and the default threshold is set to 0.01
     def __init__(self, *, alpha, threshold=1e-2, loss=sigmoid_loss):
         self.weights = []
         self.alpha = alpha
         self.threshold = threshold
         self.loss = loss
 
+    # This method is fitting the weights using the training X sample
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/gradient_descent.py]: This method fits the weights using the training sample X.

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/gradient_descent.py [hunk 6]
```diff
@@ -3,30 +3,50 @@
...
+        # The algorithm will be interrupted after reaching the threshold
         while True:
             M = X.dot(self.weights) * y
-            loss, derivative = self.loss(M)
+            loss, derivative =  self.loss(M)
 
-            grad_q = np.sum((derivative.T * (X.T * y)).T, axis=0)
+            # The gradient is a sum along the 0 axis of the following array:
+            # transposed X multiplied by y (X.T * y)
+            # then multiplied by transposed derivative (derivative.T)
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/gradient_descent.py]: multiplied by the transposed derivative (derivative.T)

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/gradient_descent.py [hunk 7]
```diff
@@ -3,30 +3,50 @@
...
                 break
-            self.weights = new_weights
+            # Update the stored weights before the next iteration
+            self.weights =  new_weights
+        # The method returns the losses stored along the fitting process
         return errors
 
     def predict(self, X):
+        # As we have only two normalized classes we can return the np.sign of the
+        # X dot product weights
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/gradient_descent.py]: dot product between X and the weights.

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/stochastic_gradient_descent.py [hunk 8]
```diff
@@ -23,20 +23,31 @@ def fit(self, X, y):
...
+            # Updating the weights
             self.weights -= self.alpha * grad
 
         return errors
 
     def calc_grad(self, w, X, y):
-        M = X.dot(w) * y
-        loss, loss_deric = self.loss(M)
-        grad = X.T * loss_deric * y
+        # margin indicates how "deep" is an object inside its class. The less the margin
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/stochastic_gradient_descent.py]: The margin indicates how "deep" an object is within its class. The smaller the margin,

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py [hunk 9]
```diff
@@ -2,12 +2,29 @@
 from utils import plot_classification
 
 
+# The function will load data from a text file with a relative path
+# 'fname'
+# and return it as a pair of arrays - one for different features
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py]: and returns it as a pair of arrays: features

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py [hunk 10]
```diff
@@ -2,12 +2,29 @@
...
+# and the other one for a diabetes presence
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py]: and diabetes presence.

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py [hunk 11]
```diff
@@ -2,12 +2,29 @@
...
 def read_data(fname):
+    # genfromtxt method loads data from a text file, using a delimeter
+    # to distinguish the columns
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py]: based on the provided delimiter.

File: Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py [hunk 12]
```diff
@@ -2,12 +2,29 @@
...
     data = np.genfromtxt(fname, delimiter=',')
+    # The data is split into the all but the last column for X and
+    # the last one for y respectively
     X, y = data[:, :-1], data[:, -1]
-    X = (X - X.mean(axis=0)) / X.std(axis=0)  # rescale features
+    # The features are rescaled:
+    # X is standardized - all values are centered around mean
+    # with a unit standard deviation. It means if we will calculate mean
+    # and standard deviation of standard scores it will be 0 and 1 respectively.
+    # This procedure is recommended if the data follows normal distribution
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Gradient Descent/task.py]: This procedure is recommended for data that follows a normal distribution.

File: Pima Indians Diabetes and Linear Classifier/Loss function/gradient_descent.py [hunk 13]
```diff
@@ -3,30 +3,50 @@
...
 class GradientDescent:
+    # The weighs will be randomly initiated as the first step of the fitting process.
+    # The default loss function is sigmoid, and the default threshold is set to 0.01
     def __init__(self, *, alpha, threshold=1e-2, loss=sigmoid_loss):
         self.weights = []
         self.alpha = alpha
         self.threshold = threshold
         self.loss = loss
 
+    # This method is fitting the weights using the training X sample
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/gradient_descent.py]: This method fits the weights using the training sample X.

File: Pima Indians Diabetes and Linear Classifier/Loss function/gradient_descent.py [hunk 14]
```diff
@@ -3,30 +3,50 @@
...
+        # The algorithm will be interrupted after reaching the threshold
         while True:
-            M = X.dot(self.weights) * y
-            loss, derivative = self.loss(M)
+            M = # Set M to a dot product of X and weights multiplied by y
+            loss, derivative = # calculate them using loss function of the instance
 
-            grad_q = np.sum((derivative.T * (X.T * y)).T, axis=0)
-            new_weights = self.weights - self.alpha * grad_q
+            # The gradient is a sum along the 0 axis of the following array:
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/gradient_descent.py]: The gradient is the sum along axis 0 of the following array:

File: Pima Indians Diabetes and Linear Classifier/Loss function/stochastic_gradient_descent.py [hunk 15]
```diff
@@ -23,20 +23,31 @@ def fit(self, X, y):
         q, _ = self.calc_grad(self.weights, X, y)
         # eta value should be in [0,1] range, this one is chosen arbitrary
         eta = 1 / len(y)
+        # the algorithm will iterate exactly n_iter times
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/stochastic_gradient_descent.py]: The algorithm will iterate exactly n_iter times

File: Pima Indians Diabetes and Linear Classifier/Loss function/stochastic_gradient_descent.py [hunk 16]
```diff
@@ -23,20 +23,31 @@ def fit(self, X, y):
...
 
     def calc_grad(self, w, X, y):
-        M = X.dot(w) * y
-        loss, loss_deriv = self.loss(M)
-        grad = X.T * loss_deriv * y
+        # margin indicates how "deep" is an object inside its class. The less the margin
+        # the closer it is to the class boundary on the hyperplane
+        margin = X.dot(w) * y
+        loss, loss_derivative = self.loss(margin)
+        # The gradient could be evaluated as a multiplication product of
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/stochastic_gradient_descent.py]: The gradient is computed as the product of

File: Pima Indians Diabetes and Linear Classifier/Loss function/task.py [hunk 17]
```diff
@@ -1,16 +1,31 @@
-import matplotlib.pyplot
 import numpy as np
 
 
+# The function will load data from a text file with a relative path
+# 'fname'
+# and return it as a pair of arrays - one for different features
+# and the other one for a diabetes presence
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/task.py]: and diabetes presence.

File: Pima Indians Diabetes and Linear Classifier/Loss function/task.py [hunk 18]
```diff
@@ -1,16 +1,31 @@
...
 def read_data(fname):
+    # genfromtxt method loads data from a text file, using a delimeter
+    # to distinguish the columns
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/task.py]: based on the provided delimiter.

File: Pima Indians Diabetes and Linear Classifier/Loss function/task.py [hunk 19]
```diff
@@ -1,16 +1,31 @@
...
     data = np.genfromtxt(fname, delimiter=',')
+    # The data is split into the all but the last column for X and
+    # the last one for y respectively
     X, y = data[:, :-1], data[:, -1]
-    X = (X - X.mean(axis=0)) / X.std(axis=0)  # rescale features
+    # The features are rescaled:
+    # X is standardized - all values are centered around mean
+    # with a unit standard deviation. It means if we will calculate mean
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/task.py]: with a unit standard deviation. This means that the mean

File: Pima Indians Diabetes and Linear Classifier/Loss function/task.py [hunk 20]
```diff
@@ -1,16 +1,31 @@
...
+    # and standard deviation of standard scores it will be 0 and 1 respectively.
+    # This procedure is recommended if the data follows normal distribution
+    X = (X - X.mean(axis=0)) / X.std(axis=0)
+    # A column with '-1' values is added to the left of the X array
+    # It is a pseudo-feature that will allow us to build initial vectors
+    # for all of the cases. It will be essential later
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Loss function/task.py]: calculations later on.

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py [hunk 21]
```diff
@@ -23,6 +23,20 @@ def fit(self, X, y):
         q, _ = self.calc_grad(self.weights, X, y)
         # eta value should be in [0,1] range, this one is chosen arbitrary
         eta = 1 / len(y)
+
+        # The algorithm will iterate exactly n_iter times
+        for i in range(self.n_iter):
+            # We need to generate a batch of random k indices from the sample
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py]: We need to generate a batch of random k indices from the sample.

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py [hunk 22]
```diff
@@ -23,6 +23,20 @@ def fit(self, X, y):
...
+            # np.random.choice is a perfect fit for this
+            batch_index = np.random.choice(len(y), self.k)
+            loss, grad = self.calc_grad(self.weights, X[batch_index], y[batch_index])
+            # Here we evaluate the quality functional
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py]: Here, we evaluate the quality functional.

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py [hunk 23]
```diff
@@ -34,9 +48,15 @@ def fit(self, X, y):
         return errors
 
     def calc_grad(self, w, X, y):
-        M = X.dot(w) * y
-        loss, loss_deric = self.loss(M)
-        grad = X.T * loss_deric * y
+        # Margin indicates how "deep" is an object inside its class. The less the margin
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/stochastic_gradient_descent.py]: The margin indicates how "deep" an object is within its class. The smaller the margin,

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py [hunk 24]
```diff
@@ -2,16 +2,32 @@
 from utils import plot_classification
 
 
+# The function will load data from a text file with a relative path
+# 'fname'
+# and return it as a pair of arrays - one for different features
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py]: and returns it as a pair of arrays: features

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py [hunk 25]
```diff
@@ -2,16 +2,32 @@
...
+# and the other one for a diabetes presence
 def read_data(fname):
+    # genfromtxt method loads data from a text file, using a delimeter
+    # to distinguish the columns
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py]: based on the provided delimiter.

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py [hunk 26]
```diff
@@ -2,16 +2,32 @@
...
     data = np.genfromtxt(fname, delimiter=',')
+    # The data is split into the all but the last column for X and
+    # the last one for y respectively
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py]: y (the last column).

File: Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py [hunk 27]
```diff
@@ -2,16 +2,32 @@
...
     X, y = data[:, :-1], data[:, -1]
-    X = (X - X.mean(axis=0)) / X.std(axis=0)  # rescale features
+    # The features are rescaled:
+    # X is standardized - all values are centered around mean
```
Comment (stephen-hero) [inline Pima Indians Diabetes and Linear Classifier/Stochastic Gradient Descent/task.py]: X is standardized by centering features around the mean
